library/alloc/src/boxed.rs

```diff
@@ -1016,6 +1016,177 @@ impl<T: ?Sized, A: Allocator> Box<T, A> {
+    /// let mut numbers: Box<[i32]> = Box::new([1, 2, 3]);
+    ///
+    /// // This will not reallocate.
+    /// Box::set(&mut numbers, [4, 5, 6]);
+    /// assert_eq!(&*numbers, &[4, 5, 6]);
+    ///
+    /// // This will reallocate.
+    /// Box::set(&mut numbers, [4, 5, 6, 7]);
+    /// assert_eq!(&*numbers, &[4, 5, 6, 7]);
+    /// ```
+    #[unstable(feature = "box_reuse", issue = "none")]
+    #[cfg(not(no_global_oom_handling))]
+    pub fn set<U>(b: &mut Box<T, A>, val: U)
+    where
+        U: Unsize<T>,
+    {
+        let layout = Layout::new::<U>();
+        if Box::try_set(b, val).is_err() {
+            handle_alloc_error(layout);
+        }
+    }
+
+    /// Sets the value inside the box, reusing the existing storage if possible and reallocating if
+    /// not, returning an error if reallocation failed.
+    ///
+    /// If the old and new values stored in the box have the same [memory layout](Layout), then the
+    /// contents of the `Box` will be updated in-place and no memory allocation will occur.
+    /// Otherwise, a new `Box` will be created to replace the old one, causing a memory allocation.
+    /// To instead return an error if this happens, use [`Box::reuse`].
+    ///
+    /// # Examples
+    ///
+    /// Reuse a boxed slice:
+    ///
+    /// ```
+    /// #![feature(allocator_api, box_reuse)]
+    ///
+    /// let mut numbers: Box<[i32]> = Box::try_new([1, 2, 3])?;
+    ///
+    /// // This will not reallocate.
+    /// Box::try_set(&mut numbers, [4, 5, 6])?;
+    /// assert_eq!(&*numbers, &[4, 5, 6]);
+    ///
+    /// // This will reallocate.
+    /// Box::try_set(&mut numbers, [4, 5, 6, 7])?;
+    /// assert_eq!(&*numbers, &[4, 5, 6, 7]);
+    /// # Ok::<(), std::alloc::AllocError>(())
+    /// ```
+    #[unstable(feature = "box_reuse", issue = "none")]
+    // #[unstable(feature = "allocator_api", issue = "32838")]
+    pub fn try_set<U>(b: &mut Box<T, A>, val: U) -> Result<(), AllocError>
+    where
+        U: Unsize<T>,
+    {
+        if let Err(val) = Box::reuse(b, val) {
+            // Due to https://github.com/rust-lang/rust/issues/78459, it is not currently possible
+            // to use a `Box<T, &A>`, so we have to work around it.
+
+            // Allocate the replacement box now so we don't have to deal with code panicking later.
+            let alloc = Box::allocator(b);
+            // Use allocation methods directly instead of `Box::try_new_in` so ownership of the
+            // allocator is not lost if allocation errors or panics.
+            let ptr: *mut U = alloc.allocate(Layout::new::<U>())?.cast().as_ptr();
+            unsafe { ptr::write(ptr, val) };
+            let new_b = unsafe { Box::from_raw_in(ptr, ptr::read(alloc)) };
+
+            let old_b = mem::replace(b, new_b);
+
+            // Drop the old box without dropping its allocator, as the allocator has been moved
+            // into the new box.
+            struct DeallocGuard<A: Allocator> {
+                alloc: A,
```

Comment:
```suggestion
            struct DeallocGuard<'alloc, A: Allocator> {
                alloc: &'alloc A,
```

to make it clear that the field `alloc` has no drop glue attached to it. Another option would be to have `alloc: ManuallyDrop<A>`, but it seems overkill here.